akka-http-core/src/main/scala/akka/http/impl/engine/http2/Http2Demux.scala

```diff
@@ -357,25 +360,50 @@ private[http2] abstract class Http2Demux(http2Settings: Http2CommonSettings, ini
         }
       })
 
+      // -----------------------------------------------------------------
       // FIXME: What if user handler doesn't pull in new substreams? Should we reject them
-      //        after a while or buffer only a limited amount? We should also be able to
-      //        keep the buffer limited to the number of concurrent streams as negotiated
-      //        with the other side.
+      //        after a while or buffer only a limited amount?
       val bufferedSubStreamOutput = new BufferedOutlet[Http2SubStream](substreamOut)
       override def dispatchSubstream(initialHeaders: ParsedHeadersFrame, data: Source[Any, Any], correlationAttributes: Map[AttributeKey[_], _]): Unit =
         bufferedSubStreamOutput.push(Http2SubStream(initialHeaders, data, correlationAttributes))
 
+      // -----------------------------------------------------------------
+      private var completing = false
+      private def tryComplete(): Unit = {
+        completing = true
+        if (activeStreamCount() == 0) complete()
+      }
+      override def onAllStreamsClosed(): Unit = if (completing) complete()
+      // Customized replacement for completeStage()
+      private def complete(): Unit = {
+        cancel(substreamIn)
+        cancel(frameIn)
+        complete(frameOut)
+        // Unlike the default `completeStage`, we need to delegate to
+        // bufferedSubStreamOutput#complete(substreamOut) instead of complete(substreamOut)
+        bufferedSubStreamOutput.complete()
+      }
+
+      // -----------------------------------------------------------------
       setHandler(substreamIn, new InHandler {
         def onPush(): Unit = {
           val sub = grab(substreamIn)
           handleOutgoingCreated(sub)
           // Once the incoming stream is handled, we decide if we need to pull more.
           tryPullSubStreams()
         }
+
+        override def onUpstreamFinish(): Unit = {
+          // marks StreamHandling as ready for completion
+          tryComplete()
```

Comment:
Do we still need `tryComplete`?

Reply:
see https://github.com/akka/akka-http/pull/3734#discussion_r561190431